src/libcore/iter/adapters/zip.rs

```diff
@@ -289,6 +293,31 @@ where
 {
 }
 
+// Arbitrarily selects the left side of the zip iteration as extractable "source"
+// it would require negative trait bounds to be able to try both
+#[unstable(issue = "none", feature = "inplace_iteration")]
+unsafe impl<S, A, B> SourceIter for Zip<A, B>
+where
+    A: SourceIter<Source = S>,
+    B: Iterator,
+    S: Iterator,
+{
+    type Source = S;
+
+    #[inline]
+    unsafe fn as_inner(&mut self) -> &mut S {
+        SourceIter::as_inner(&mut self.a)
+    }
+}
+
+#[unstable(issue = "none", feature = "inplace_iteration")]
+// Limited to Item: Copy since interaction between Zip's use of TrustedRandomAccess
```

Comment:
How does TrustedRandomAccess relate to the specialization InPlaceIterable is used for?

Reply:
Zip + in-place iteration performed worse than a for loop + direct in-place mutation. Making `vec::IntoIter` implement `TrustedRandomAccess` enabled the use of Zip's specialization on TRA. But the interaction between TRA not calling `next()`, in-place-iteration and getting head/tail drop right seemed to complex, so I simply say that in place iteration *for Zip adapters* only makes sense when the type that we want to do the in-place stuff on doesn't need drop.